internal/appsec/limiter.go

```diff
@@ -0,0 +1,96 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2022 Datadog, Inc.
+
+//go:build appsec
+// +build appsec
+
+package appsec
+
+import (
+	"sync/atomic"
+	"time"
+)
+
+type Limiter interface {
+	Allow() bool
+}
+
+//The token ticker is a thread-safe and lock-free rate limiter based on a token bucket.
+//The idea is to have a goroutine that will update  the bucket with fresh tokens at regular intervals using a time.Ticker.
+//The advantage of using a goroutine here is  that the implementation becomes easily thread-safe using a few
+//atomic operations with little overhead overall. TokenTicker.Start() *must* be called before the first call to
+//TokenTicker.Allow() and TokenTicker.Stop() *must* be called once done using.
+type TokenTicker struct {
+	tokens    int64
+	maxTokens int64
+	ticker    *time.Ticker
+	stopChan  chan bool
+}
+
+func NewTokenTicker(tokens, maxTokens int64) *TokenTicker {
+	return &TokenTicker{
+		tokens:    tokens,
+		maxTokens: maxTokens,
+		stopChan:  make(chan bool),
+	}
+}
+
+func (t *TokenTicker) Start() {
+	if t.ticker != nil {
+		t.Stop()
+	}
+
+	t.ticker = time.NewTicker(500 * time.Microsecond)
+	//Ticker goroutine: ticks every 500ms to check whether tokens can be added to the bucket or not
+	go func(t *TokenTicker) {
+		ticksPerToken := (time.Second.Nanoseconds() / t.maxTokens)
+		ticks := int64(0)
+		prevStamp := time.Now()
+
+		for {
+			select {
+			case <-t.stopChan:
+				return
+			case stamp := <-t.ticker.C:
+				ticks += stamp.Sub(prevStamp).Nanoseconds()
+				prevStamp = stamp
+				if ticks >= ticksPerToken {
+					for {
+						tokens := atomic.LoadInt64(&t.tokens)
+						if tokens == t.maxTokens {
+							break
+						}
+						inc := ticks / ticksPerToken
+						if tokens+inc > t.maxTokens {
+							inc -= (tokens + inc) % t.maxTokens
+						}
+						if atomic.CompareAndSwapInt64(&t.tokens, tokens, tokens+inc) {
+							ticks = ticks % ticksPerToken
+							break
+						}
+					}
+				}
+			}
+		}
+	}(t)
+}
+
+func (t *TokenTicker) Stop() {
```

Comment:
Would you consider using the already existing code for the rate limiter? https://github.com/DataDog/dd-trace-go/blob/2ed47cc50e51f71fd5929f3ca8eca56d850eedc2/ddtrace/tracer/sampler.go#L412

It could be factored out into an `internal` package. Why re-invent the wheel? Furthermore, even if you don't use that one specifically, the project already imports [golang.org/x/time/rate](https://github.com/DataDog/dd-trace-go/blob/2ed47cc50e51f71fd5929f3ca8eca56d850eedc2/ddtrace/tracer/sampler.go#L24), so why not use that instead?

Reply:
> Would you consider using the already existing code for the rate limiter?

That's the goal in the future. We want to introduce this implementation because it was the easiest/quickest way for us to add lock-free rate limiting. Some comments about this have been made on the RFC, [here](https://docs.google.com/document/d/1X64XQOk3N-aS_F0bJuZLkUiJqlYneDxo_b8WnkfFy_0/edit).

We agreed with Raphael that a good compromise would be to re-use the APM code if it is possible to move it to a lock-free version. This feature was added kinda last minute for 1.36.0 so this would allow us to implement it while we move on to a solution that's better for everyone for GA.

Reply:
Got it. Thanks for providing the detail.